crates/matrix-sdk-store-encryption/src/lib.rs

```diff
@@ -0,0 +1,536 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+// Copyright 2021 Damir Jelić
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![doc = include_str!("../README.md")]
+#![warn(missing_debug_implementations, missing_docs)]
+
+use blake3::{derive_key, Hash};
+use chacha20poly1305::{
+    aead::{Aead, Error as EncryptionError, NewAead},
+    Key as ChachaKey, XChaCha20Poly1305, XNonce,
+};
+use displaydoc::Display;
+use hmac::Hmac;
+use pbkdf2::pbkdf2;
+use rand::{thread_rng, Error as RandomError, Fill};
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use zeroize::Zeroize;
+
+const VERSION: u8 = 1;
+const KDF_SALT_SIZE: usize = 32;
+const XNONCE_SIZE: usize = 24;
+#[cfg(not(test))]
+const KDF_ROUNDS: u32 = 200_000;
+#[cfg(test)]
+const KDF_ROUNDS: u32 = 1000;
+
+type MacKeySeed = [u8; 32];
+
+/// Error type for the `StoreCipher` operations.
+#[derive(Debug, Display, thiserror::Error)]
+pub enum Error {
+    /// Failed to serialize or deserialize a value {0}
+    Serialization(#[from] serde_json::Error),
+    /// Error encrypting or decrypting a value {0}
+    Encryption(#[from] EncryptionError),
+    /// Coulnd't generate enough randomness for a cryptographic operation: {0}
+    Random(#[from] RandomError),
+    /// Unsupported ciphertext version, expected {0}, got {1}
+    Version(u8, u8),
+    /// The ciphertext had an invalid length, expected {0}, got {1}
+    Length(usize, usize),
+}
+
+/// An encryption key that can be used to encrypt data for key/value stores.
+///
+/// # Examples
+///
+/// ```
+/// # let example = || {
+/// use matrix_sdk_store_encryption::StoreCipher;
+/// use serde_json::{json, value::Value};
+///
+/// let store_cipher = StoreCipher::new()?;
+///
+/// // Export the store key and persist it in your key/value store
```

Comment:
Seems we missed the phrase `store key`, so please also run `s/store key/store cipher/g`.